tests/roles/pcp_cleanup/tasks/main.yaml

```diff
@@ -2,29 +2,49 @@
   ansible.builtin.shell: |
     {{ shell_header }}
     {{ oc_header }}
+    # Cleanup OpenStackControlPlane object
+    oc delete --ignore-not-found=true osctlplane -l core.openstack.org/openstackcontrolplane -n openstack
 
-    oc delete --ignore-not-found=true --wait=false openstackdataplanedeployment/openstack
-    oc delete --ignore-not-found=true --wait=false openstackdataplanedeployment/openstack-nova-compute-ffu
-    oc delete --ignore-not-found=true --wait=false openstackcontrolplane/openstack
-    oc patch openstackcontrolplane openstack --type=merge --patch '
-    metadata:
-      finalizers: []
-    ' || true
-
-    while oc get pod | grep rabbitmq-server-0; do
-        sleep 2
-    done
-    while oc get pod | grep openstack-galera-0; do
-        sleep 2
+    # Ensure that all pods in openstack namespace are deleted
+    while oc get pod -n openstack | grep -E 'rabbitmq-server-0|openstack-galera-0'; do
+      sleep 2;
     done
 
+    # Delete Control Plane CRD
+    oc delete crd --ignore-not-found=true openstackcontrolplanes.core.openstack.org
+
+    # Cleanup OpenStackDataplane objects
+    oc delete --ignore-not-found=true OpenStackDataPlaneDeployment --all -n openstack
+    oc delete --ignore-not-found=true OpenStackDataPlaneNodeSet --all -n openstack
+    oc delete --ignore-not-found=true OpenStackDataPlaneService --all -n openstack
+
+    # Delete Dataplane CRDs
+    oc delete crd --ignore-not-found=true openstackdataplanedeployments.dataplane.openstack.org
+    oc delete crd --ignore-not-found=true openstackdataplanenodesets.dataplane.openstack.org
+    oc delete crd --ignore-not-found=true openstackdataplaneservices.dataplane.openstack.org
+
```

Comment:
what is meant with between runs? between rerun of a full update job? 

before we deploy the initial environment to which should get updated, we probably want to delete the CRDs from a previous run and it might be required, e.g. in the case where we do an update to an operator which introduces v1beta2. If you start with an operator version which only supports v1beta1 you do not want to have a v1beta2 from an older job installed. I think this version would not be removed when you install the operator which won't support it.

When you update the operators for a minor update you can not delete the CRDs. if you delete CRDs the CRs get deleted as well and your deployment is gone, yes a finalizer might block it. The updated operator will handle installing CRD changes for new versions like @gibizer said.

Reply:
"Between runs" means between 2 CI runs... It's full update job run eventually.

Our job doesn't deploy OCP on the run as developers crc. It reuses static OCP that is deployed once outside of the jobs. When CI is triggered, it will deploy OSP platform and will deploy operators on provided OCP. When job fails it means that all operators, crds/crs, csvs, secrets are left. When clean up is not perform the next CI run will have a failure. When I debug it I got the following situation

```
oc describe subscription openstack-operator

    Message:                error updating CRD "instancehas.instanceha.openstack.org": CustomResourceDefinition.apiextensions.k8s.io "instancehas.instanceha.openstack.org" is invalid: spec.names.kind: Invalid value: "InstanceHa": field is immutable
    Reason:                 InstallComponentFailed
```
I was able to fix it by running 
```
oc delete crd instancehas.instanceha.openstack.org
```
and triggering CI again.

I understand minor update case. However, the job I am working on is not minor update. It's adoption job. The job deploys OSP 17.1 and tries to adopt environment to 18.0 upstream master content. If crd structure is changed the operator will fail as it won't be able to patch. See error above.